File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/lovvalg/A003MapperTest.java
package no.nav.melosys.eessi.service.sed.mapper.til_sed.lovvalg;

import java.io.IOException;
import java.net.URISyntaxException;
import java.time.LocalDate;

import no.nav.melosys.eessi.controller.dto.SedDataDto;
import no.nav.melosys.eessi.controller.dto.VedtakDto;
import no.nav.melosys.eessi.models.exception.MappingException;
import no.nav.melosys.eessi.models.exception.NotFoundException;
import no.nav.melosys.eessi.models.sed.SED;
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA003;
import no.nav.melosys.eessi.service.sed.SedDataStub;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class A003MapperTest {

    private final LovvalgSedMapper<MedlemskapA003> sedMapper = new A003Mapper();
    private SedDataDto sedData;

    @BeforeEach
    public void setup() throws IOException, URISyntaxException {
        sedData = SedDataStub.getStub();
        sedData.getLovvalgsperioder().get(0).setLovvalgsland("NO");

    }

    @Test
    void mapTilSed() throws MappingException, NotFoundException {
        SED sed = sedMapper.mapTilSed(sedData);

        assertThat(sed).isNotNull();
        assertThat(sed.getMedlemskap()).isInstanceOf(MedlemskapA003.class);

        MedlemskapA003 medlemskapA003 = (MedlemskapA003) sed.getMedlemskap();
        assertThat(sed.getNav().getArbeidsgiver()).allMatch(a -> "NO".equals(a.getAdresse().getLand()));
        assertThat(medlemskapA003.getAndreland().getArbeidsgiver()).noneMatch(a -> "NO".equals(a.getAdresse().getLand()));
    }

    @Test
    void erIkkeOpprinneligVedtak_ErOpprinneligVedtakOgErEndringsvedtakSattKorrektOgDatoForrigeVedtakIkkeNull() {
        VedtakDto vedtakDto = new VedtakDto();
        vedtakDto.setErFørstegangsvedtak(false);
        vedtakDto.setDatoForrigeVedtak(LocalDate.now());
        sedData.setVedtakDto(vedtakDto);


        SED sed = sedMapper.mapTilSed(sedData);

        assertThat(sed.getMedlemskap().getClass()).isEqualTo(MedlemskapA003.class);
        MedlemskapA003 medlemskapA003 = (MedlemskapA003) sed.getMedlemskap();
        assertThat(medlemskapA003).isNotNull();
        assertThat(medlemskapA003.getVedtak().getEropprinneligvedtak()).isNull();
        assertThat(medlemskapA003.getVedtak().getErendringsvedtak()).isEqualTo("nei");
        assertThat(medlemskapA003.getVedtak().getDatoforrigevedtak()).isNotNull();
        assertThat(medlemskapA003.getVedtak().getDatoforrigevedtak()).isEqualTo(LocalDate.now().toString());
    }

    @Test
    void erOpprinneligVedtak_ErOpprinneligVedtakOgErEndringsvedtakSattKorrektOgDatoForrigeVedtakNull() {
        VedtakDto vedtakDto = new VedtakDto();
        vedtakDto.setErFørstegangsvedtak(true);
        sedData.setVedtakDto(vedtakDto);


        SED sed = sedMapper.mapTilSed(sedData);

        assertThat(sed.getMedlemskap().getClass()).isEqualTo(MedlemskapA003.class);
        MedlemskapA003 medlemskapA003 = (MedlemskapA003) sed.getMedlemskap();
        assertThat(medlemskapA003).isNotNull();
        assertThat(medlemskapA003.getVedtak().getEropprinneligvedtak()).isEqualTo("ja");
        assertThat(medlemskapA003.getVedtak().getErendringsvedtak()).isNull();
        assertThat(medlemskapA003.getVedtak().getDatoforrigevedtak()).isNull();
    }
}

File: melosys-eessi-app/src/test/java/no/nav/melosys/eessi/service/sed/mapper/til_sed/lovvalg/A003MapperTest.kt
package no.nav.melosys.eessi.service.sed.mapper.til_sed.lovvalg

import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.types.shouldBeInstanceOf
import no.nav.melosys.eessi.controller.dto.SedDataDto
import no.nav.melosys.eessi.controller.dto.VedtakDto
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.models.sed.medlemskap.impl.MedlemskapA003
import no.nav.melosys.eessi.service.sed.SedDataStub
import org.junit.jupiter.api.Test
import java.time.LocalDate

class A003MapperTest {
    private fun mapTilA003(
        erCDM4_3: Boolean,
        block: SedDataDto.() -> Unit = {}
    ): SED =
        SedDataStub.mapTilSed<A003Mapper>(erCDM4_3, "mock/sedDataDtoStub.json") {
            lovvalgsperioder.first().lovvalgsland = "NO"
            block()
        }

    @Test
    fun `map til SED med version 2`() {
        val sed = mapTilA003(false)

        sed.shouldNotBeNull().run {
            medlemskap.shouldBeInstanceOf<MedlemskapA003>().run {
                andreland.shouldNotBeNull()
                    .arbeidsgiver.shouldNotBeNull().single()
                    .adresse.shouldNotBeNull()
                    .land shouldNotBe "NO"
            }
            nav.shouldNotBeNull().run {
                arbeidsgiver.shouldNotBeNull().single().adresse.shouldNotBeNull().land shouldBe "NO"
                arbeidsland shouldBe null
            }
            sedVer shouldBe "2"
            sedGVer shouldBe "4"
        }
    }

    @Test
    fun `map til SED version 3`() {
        val sed = mapTilA003(true)

        sed.shouldNotBeNull().run {
            medlemskap.shouldBeInstanceOf<MedlemskapA003>().run {
                andreland.shouldNotBeNull()
                    .arbeidsgiver.shouldNotBeNull().single().adresse.shouldNotBeNull().land shouldNotBe "NO"
            }
            nav.shouldNotBeNull().run {
                arbeidsgiver.shouldNotBeNull().single().adresse.shouldNotBeNull().land shouldBe "NO"
                arbeidsland.shouldNotBeNull().size shouldBe 1
            }
            sedVer shouldBe "3"
            sedGVer shouldBe "4"
        }
    }

    @Test
    fun `er ikke opprinnelig vedtak forvent korrekt verdier`() {
        val sed = mapTilA003(false) {
            vedtakDto = VedtakDto(
                erFørstegangsvedtak = false,
                datoForrigeVedtak = LocalDate.now()
            )
        }

        sed.medlemskap.shouldBeInstanceOf<MedlemskapA003>().run {
            vedtak.shouldNotBeNull().run {
                eropprinneligvedtak shouldBe null
                erendringsvedtak shouldBe "nei"
                datoforrigevedtak.shouldNotBeNull() shouldBe LocalDate.now().toString()
            }
        }
    }

    @Test
    fun `er opprinnelig vedtak forvent korrekt verdier`() {
        val sed = mapTilA003(false) {
            vedtakDto = VedtakDto(erFørstegangsvedtak = true)
        }

        sed.medlemskap.shouldBeInstanceOf<MedlemskapA003>().run {
            vedtak.shouldNotBeNull().run {
                eropprinneligvedtak shouldBe "ja"
                erendringsvedtak shouldBe null
                datoforrigevedtak shouldBe null
            }
        }
    }
}
